scrapy/dispatch/dispatcher.py

```diff
@@ -0,0 +1,367 @@
+                anything hashable.
+        """
+
+        # If DEBUG is on, check that we got a good receiver
+        # if settings.DEBUG:
+        assert callable(receiver), "Signal receivers must be callable."
+
+        # Check for **kwargs
+        if not func_accepts_kwargs(receiver):
+            raise ValueError(
+                "Signal receivers must accept keyword arguments (**kwargs).")
+
+        if dispatch_uid:
+            lookup_key = (dispatch_uid, _make_id(sender))
+        else:
+            lookup_key = (_make_id(receiver), _make_id(sender))
+
+        if weak:
+            ref = weakref.ref
+            receiver_object = receiver
+            # Check for bound methods
+            if hasattr(receiver, '__self__') and hasattr(receiver, '__func__'):
+                ref = WeakMethod
+                receiver_object = receiver.__self__
+            if six.PY3 and sys.version_info >= (3, 4):
+                receiver = ref(receiver)
+                weakref.finalize(receiver_object, self._remove_receiver)
+            else:
+                receiver = ref(receiver, self._remove_receiver)
+
+        with self.lock:
+            self._clear_dead_receivers()
+            for r_key, _ in self.receivers:
+                if r_key == lookup_key:
+                    break
+            else:
+                self.receivers.append((lookup_key, receiver))
+            self.sender_receivers_cache.clear()
+
+    def disconnect(self, receiver=None, sender=None, weak=None, dispatch_uid=None):
+        """
+        Disconnect receiver from sender for signal.
+
+        If weak references are used, disconnect need not be called. The receiver
+        will be remove from dispatch automatically.
+
+        Arguments:
+
+            receiver
+                The registered receiver to disconnect. May be none if
+                dispatch_uid is specified.
+
+            sender
+                The registered sender to disconnect
+
+            dispatch_uid
+                the unique identifier of the receiver to disconnect
+        """
+        if weak is not None:
+            warnings.warn("Passing `weak` to disconnect has no effect.",
+                          ScrapyDeprecationWarning, stacklevel=2)
+        if dispatch_uid:
+            lookup_key = (dispatch_uid, _make_id(sender))
+        else:
+            lookup_key = (_make_id(receiver), _make_id(sender))
+
+        disconnected = False
+        with self.lock:
+            self._clear_dead_receivers()
+            for index in range(len(self.receivers)):
+                (r_key, _) = self.receivers[index]
+                if r_key == lookup_key:
+                    disconnected = True
+                    del self.receivers[index]
+                    break
+            self.sender_receivers_cache.clear()
+        return disconnected
+
+    def has_listeners(self, sender=None):
+        return bool(self._live_receivers(sender))
+
+    def send(self, sender, **named):
+        """
```

Comment:
It would be good if the docstrings were formatted in the same fashion as [other ones in `scrapy`](https://github.com/scrapy/scrapy/blob/master/scrapy/settings/__init__.py#L114), but you'll probably encounter this anyways when you do docs